### PR TITLE
alloc: fix the general one-buffer memory allocation case

### PR DESCRIPTION
**Pull request by lyakh**
**Note** this has to wait for #3642 to be merged first to then get rebased on top of it.
The condition "size + alignment <= block_size" for allocating memory from a signle buffer is sufficient but not precise enough. For example if we want to allocate 20 bytes with 64-byte alignment, a 32-byte buffer *might* be sufficient if it's suitably aligned. Fix the algorithm to account for such cases.

### CODE REVIEW

**Comment by lyakh** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
@zrombel what about this one? Also device issues?

**Comment by zrombel** [comment]
@lyakh It looks like real issue. On platforms all platforms where Keyword Detection tests apply there is a DSP panic.

**Comment by lgirdwood** [comment]
@lyakh it's worth checking the KWD here, as it may be relying on the current behaviour - even if's non optimal.

**Review by lgirdwood — APPROVED** [review]
@lyakh can you check KWD before and after this PR. It maybe obvious from trace output.

**Comment by zrombel** [comment]
This PR is causing failures on test platforms. I've temporary blacklisted it. Please let me know when PR will be ready for testing and merging.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@lyakh is this ready for testing now. please let @zrombel know so CI can run.

**Comment by lyakh** [comment]
> @lyakh is this ready for testing now. please let @zrombel know so CI can run.

@lgirdwood I was waiting for the latest CI run to complete, is it in "expected" state because @zrombel blocked it? If so, yes, please unblock, all the other tests look good.

**Comment by lgirdwood** [comment]
@zrombel any comment here, it appears that CI has been in expected state for 19hrs ?

**Comment by zrombel** [comment]
I've removed this PR from blacklist and scheduled it for build and testing. Results should be available during the day.

**Comment by lyakh** [comment]
> I've removed this PR from blacklist and scheduled it for build and testing. Results should be available during the day.

@zrombel thanks. Could you also explain a bit how that blacklisting works? PRs get re-tested by the CI only when they are updated, when a new revision is committed, right? So why is there a need to blacklist PRs?

**Comment by zrombel** [comment]
Yes, CI is triggered only when PR's are updated. This PR was updated couple of times and each time it got tested on CI platforms it caused platform failures so they needed manual restarts and causing CI failures of other PR's. I couldn't ques if this PR would be frequently updated or not so I put it on black list for sake of other PR's. Hopefully we won't  be needing blacklist again :)

**Comment by lyakh** [comment]
> Yes, CI is triggered only when PR's are updated. This PR was updated couple of times and each time it got tested on CI platforms it caused platform failures so they needed manual restarts and causing CI failures of other PR's. I couldn't ques if this PR would be frequently updated or not so I put it on black list for sake of other PR's. Hopefully we won't be needing blacklist again :)

@zrombel I see, thanks. Maybe it would be possible and make sense to find out why devices were crashing hard and maybe implement automatic recovery / restart?

**Comment by lyakh** [comment]
@zrombel sorry, confused again. Are tests still running or have they completed? They seem to report completion, but the output is very small and there are no failures there although the complete test reports failure. Could it be that you unblocked this PR only partially?

**Comment by zrombel** [comment]
This PR must be cursed, first it has been crashing our platforms and now logs were not uploaded to server ;) I've rerun it and now everything went as it should and you can see logs. There are two FAILS, both caused by DSP Panic in KdDmicD0ix test.

**Comment by zrombel** [comment]
Regarding recovery/restart procedure - CI infrastructure does not suport DUT hard reset at this point. Changes that would be needed to achieve this functionality would result in many others issues which we intentionally trying to avoid. Since issues like caused by this PR happens rather rarely I would keep CI as it is.

**Comment by lgirdwood** [comment]
Lets see if we can reduce the curse :) @lyakh can you split this PR into 3 PRs - one for each patch. It should then be simpler to see what's blocking the CI and will be able to merge the other 2. One of these patches could be uncovering a bug in the code .....

**Comment by lyakh** [comment]
@zrombel sorry, confused again. Are tests still running or have they completed? They seem to report completion, but the output is very small and there are no failures there although the complete test reports failure. Could it be that you unblocked this PR only partially?

> Lets see if we can reduce the curse :) @lyakh can you split this PR into 3 PRs - one for each patch. It should then be simpler to see what's blocking the CI and will be able to merge the other 2. One of these patches could be uncovering a bug in the code .....

@lgirdwood sorry, it is a good method in general, but in this case patches 2 and 3 are functional dummies, they are purely cosmetic / theoretical.

**Comment by lyakh** [comment]
@zrombel You also know contents of individual quickbuild tests, right? How are those keyword detection tests performed? There is also a KWD test in sof-test. I wanted to run it, but it requires a USB audio card. Is this also how respective quickbuild tests work? Can I reproduce somehow or at least get a DSP log?

**Comment by zrombel** [comment]
@lyakh PR was fully tested. The reason why FW trace logs are so short is 07_05_TestKdDmicD0ix16000Hz24b32b2ch test puts DSP in D0ix state and to achieve that trace logs have to be disabled. And that what makes KD D0ix issues so hard to debug. But from what I can see the problem here isn't KD it self. Test fails when stream is created so probably there is some problem with memory allocation. I can run KD tests without D0ix transition manually on Monday and provide you trace logs.

Regarding KD D0ix tests:
Python tests also uses external USB audio device and a DMIC injector which converts audio signal do PDM signals. Test creates proper pipelines, puts DSP in D0ix and plays audio signal on USB device. Audio signal consist 3s of silence and 3s of full amplitude sinusoid, so FW should detect signal after 3s, wake up and send notification to host. Test fails if there is no KD notification, of recorded signal is too short of has glitches.

**Comment by lgirdwood** [comment]
@aiChaoSONG are you able to help @lyakh here since you know the test ? @mengdonglin fyi - needed for v1.7

**Comment by lyakh** [comment]
> @aiChaoSONG are you able to help @lyakh here since you know the test ? @mengdonglin fyi - needed for v1.7

To reproduce this issue I'm also trying to run the keyword detection sof-test script, and that doesn't seem to run at all in my setup (see my today's internal mails)

**Comment by lyakh** [comment]
Found and fixed one bug. The result improved, but still not 100%...
1. the QB failure shows too high an SNR in a KWD test on JSL. The same test on the other two platforms, running it - WHL and TGL passes.
2. the device-test failure is a DMA failure on BSW.

I cannot directly relate these failures to the PR, but I cannot exclude causation either, especially in the latter case... More debugging needed, but I don't have access to BSW hardware.

**Comment by lgirdwood** [comment]
@lyakh looks good on internal CI - can you check the build CI

**Comment by lyakh** [comment]
> @lyakh looks good on internal CI - can you check the build CI

@lgirdwood Yes, I don't know what to do with sporadic failures. The previous version was the same only without one debug print. Now the internal CI had no failures. The fw-build failure is ".text segment too large." Presumably, my debug print crossed the border... Waiting for the device-test now.

**Comment by aiChaoSONG** [comment]
Build failure on BYT, due to .text section too big. @lyakh 
```
02:06:01 byt xcc build fail
02:06:01 [ 98%] Building C object CMakeFiles/sof.dir/src/schedule/task.c.o
02:06:01 [100%] Building C object CMakeFiles/sof.dir/src/schedule/timer_domain.c.o
02:06:01 [100%] Linking C executable sof
02:06:01 /srv/home/jenkins/xcc/install/tools/RD-2012.5-linux/XtensaTools/bin/xt-ld: sof section `.text' will not fit in region `sof_text_start'
02:06:01 CMakeFiles/sof.dir/build.make:1767: recipe for target 'sof' failed
02:06:01 make[3]: *** [sof] Error 2
02:06:01 CMakeFiles/Makefile2:427: recipe for target 'CMakeFiles/sof.dir/all' failed
02:06:01 make[2]: *** [CMakeFiles/sof.dir/all] Error 2
02:06:01 CMakeFiles/Makefile2:1694: recipe for target 'src/arch/xtensa/CMakeFiles/bin.dir/rule' failed
02:06:01 make[1]: *** [src/arch/xtensa/CMakeFiles/bin.dir/rule] Error 2
02:06:01 Makefile:736: recipe for target 'bin' failed
02:06:01 make: *** [bin] Error 2
```

**Comment by lyakh** [comment]
> Build failure on BYT, due to .text section too big. @lyakh @lgirdwood , on-device-test will not run if there is build failure currently.

@aiChaoSONG building only failed for one platform and this will block testing on all devices?..

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
@aiChaoSONG seems the same happened again - one compilation failed and all further testing is blocked

**Comment by aiChaoSONG** [comment]
@lyakh I triggered a test on this PR， check our internal report， 1062 1063， except the build failure on byt， no issue found.

**Comment by lyakh** [comment]
Current CI failures are: (1) device-test on ZGL - last 3 tests timed out, (2) travis - 2 builds failed because of docker rate-limits

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
